feat: Support Linux stap v3 probes

aapoalas opened the pull request:
This PR makes the crate capable of generating SystemTap's version 3 probes, also sometimes known as SDTs (no relation to DTrace's USDTs, obviously!). This is very nice on Linux systems that don't have any easy way of installing or building DTrace (such as I-use-ArchLinux-by-the-way). This works on all three of the mechanisms for converting D probe definitions into Rust. It also automatically generates the stap versions if `isenabled` probes for each probe.

The open questions in this PR as I see them are:
1. Is there any interest in upstreaming this in the first place?
2. Currently `test_does_it_work` and all tests similar to it do not work, as those rely on DTrace to work. On a generic Linux something similar might be doable with perhaps readelf, or perf, or one of the thousand different eBPF toolsets. What would make the most sense?
3. The methods added for generating GNU assembler argument format are fairly ugly, and I'm not at all sure if they really belong on the DataType enums or if they should go into the `stap3.rs` file.
4. What should I call the usdt-impl? `stap3`? `systemtap`? `ebpf`? `linux`? `knockoff`?
5. Do the `isenabled` probes make sense here? As I understand it, with DTrace USDT's the `isenabled` boolean actually lives in the instruction stream, whereas for SDT's the `isenabled` boolean is read from a mutable static. That sounds like it has a fairly real effect on runtime. If that is the case then it sounds like one would want to give the programmer a way to skip the `isenabled` probe when it isn't needed.

I understand that Oxide doesn't really have skin in the game for expanding the scope this crate to support non-DTrace probes. I'm hoping that you'd still find this potentially worth upstreaming (once the open questions and any issues you have with this PR are resolved): This crate seems to me to be unconditionally the superior way to inject USDTs into Rust code, and it would be great to spread the joy of tracing wider into the Rust ecosystem.

Hoping to hear from you all. Cheers.

ahl commented:
I'll respond in more detail but: really excited that you've taken this up and we are definitely interested in adding this functionality. As you say, it's not a top priority for Oxide but we'll carve out some time as we can to check this out

ahl commented on dtrace-parser/src/lib.rs:
```diff
@@ -110,6 +115,93 @@ impl Integer {
         format!("{prefix}int{}_t", self.width_to_c_str())
     }
 
+    pub fn to_asm_op(&self, i: u8) -> String {
```
is this the right place for this? I think the point of this crate is to parse DTrace files..

bnaecker replied:
I agree, this should probably be moved to the `usdt-impl` crate. All the types we need to operate on from the `dtrace-parser` crate are public, so we have access to everything we need there.

bnaecker replied:
Could we also provide a bit more documentation as to the purpose of this function? What does it actually produce and how is it used?

aapoalas replied:
Moved to its own file and added documentation.

ahl commented on usdt-impl/src/stap3.rs:
```diff
@@ -0,0 +1,184 @@
+//! The SystemTap probe version 3 of the USDT crate.
+//!
+//! Used on Linux platforms without DTrace.
```
I think we usually put this below the copyright and license stuff?

aapoalas replied:
Fixed.

ahl commented:
> 1. Is there any interest in upstreaming this in the first place?

Yes. We would love it if users of this crate could produce useful probes on Linux. This is--in part--why we eschewed the name `dtrace` and chose `usdt` instead.

> 2. Currently `test_does_it_work` and all tests similar to it do not work, as those rely on DTrace to work. On a generic Linux something similar might be doable with perhaps readelf, or perf, or one of the thousand different eBPF toolsets. What would make the most sense?

I think something in the eBPF universe would make sense? Something that a typical user might employ. Do you have suggestions?

> 3. The methods added for generating GNU assembler argument format are fairly ugly, and I'm not at all sure if they really belong on the DataType enums or if they should go into the `stap3.rs` file.

My inclination would be to keep as much of the specific code in the stap3.rs file (and other implementation-specific files such as no-linker.rs) as appropriate.

> 4. What should I call the usdt-impl? `stap3`? `systemtap`? `ebpf`? `linux`?

I think stap3 makes as much sense as anything. Can you walk through consideratwions?

> 5. Do the `isenabled` probes make sense here? As I understand it, with DTrace USDT's the `isenabled` boolean actually lives in the instruction stream, whereas for SDT's the `isenabled` boolean is read from a mutable static. That sounds like it has a fairly real effect on runtime. If that is the case then it sounds like one would want to give the programmer a way to skip the `isenabled` probe when it isn't needed.

I see: so with the stap implementation always performs a load from that mutable static? This seems.. nuts? Can I possibly be understanding this properly??

> 6. Is it intended that a Rust probe provider function defined as `&u32` gets translated into a probe taking a `u32` by value? See [this test](https://github.com/oxidecomputer/usdt/blob/master/usdt-attr-macro/src/lib.rs#L464).

Yes... I think. @bnaecker ?

> I understand that Oxide doesn't really have skin in the game for expanding the scope this crate to support non-DTrace probes. I'm hoping that you'd still find this potentially worth upstreaming (once the open questions and any issues you have with this PR are resolved): This crate seems to me to be unconditionally the superior way to inject USDTs into Rust code, and it would be great to spread the joy of tracing wider into the Rust ecosystem.

Are there other crates people use for Linux / eBPF / Stap / etc?

ahl reviewed (commented):
good progress. is there a good overview of the mechanism? Would it be appropriate to include an overview of how this thing works or links out? I can't recall what we have in linker.rs and no-linker.rs but certainly that would be warranted there as well.

bnaecker commented:
> > Is it intended that a Rust probe provider function defined as &u32 gets translated into a probe taking a u32 by value? See [this test](https://github.com/oxidecomputer/usdt/blob/master/usdt-attr-macro/src/lib.rs#L464).
>
> Yes... I think. @bnaecker ?

Yes, this is mostly the case. We do consider that probe to ultimately have a 32-bit, unsigned integer type, not a pointer. When we emit the probe macro itself, we try to borrow the input, but we then dereference the borrow anyway, passing a copy. For these native integer types, it doesn't matter so much since everything fits in a register. We take more care with strings to pass a pointer, but we also need to null-terminate the data anyway, which necessitates another copy.

Hope that answers your questions, but happy to dig more into the details if needed!

bnaecker commented on usdt-impl/src/empty.rs:
```diff
@@ -17,6 +17,7 @@
 // limitations under the License.
 
 use crate::common;
+use crate::record::emit_probe_record;
```
This looks unused, let's remove if so.

aapoalas replied:
Leftover code indeed. Fixed.

bnaecker commented on usdt-impl/src/common.rs:
```diff
@@ -117,26 +117,28 @@ fn shared_slice_elem_type(reference: &syn::TypeReference) -> Option<&syn::Type>
-    let abi_regs = ["x0", "x1", "x2", "x3", "x4", "x5"];
-    #[cfg(not(any(target_arch = "aarch64", target_arch = "x86_64")))]
-    compile_error!("USDT only supports x86_64 and ARM64 architectures");
+    let abi_regs = &ABI_REGS;
```
I think we can just inline this into the `.zip()` call, as `&ABI_REGS`.

aapoalas replied:
Leftover code from when I originally reused the ABI_REGS in the argument handling. Reverted these changes.

bnaecker commented on dtrace-parser/src/lib.rs:
```diff
@@ -110,6 +115,93 @@ impl Integer {
+                #[cfg(not(any(target_pointer_width = "32", target_pointer_width = "64")))]
+                BitWidth::Pointer => compile_error!("Unsupported pointer width"),
+            };
+            let suffix = match self.width {
```
I think it would be more clear to merge the matches for computing both `prefix` and `suffix`. Right now, the conditional logic on L147 is a bit obtuse, and I think the duplication is worth being very explicit about what prefix / suffix is used when.

aapoalas replied:
Merged into a relative huge match statement. Definitely clearer that way: I found that I had bugged the r8 and r9 1 byte, 2 byte and 3 byte names.

bnaecker commented on usdt-impl/src/stap3.rs:
```diff
@@ -0,0 +1,184 @@
+//!
+//! Used on Linux platforms without DTrace.
+
+// Copyright 2021 Oxide Computer Company
```
I think this is probably copy-paste. I'm actually not sure _who_ owns the copyright here (probably you, but IANAL), but the year should be current in any case.

aapoalas replied:
A lot of this file is copied over from empty and non-linker, so I kept this as Oxide Computer. The argument handling code I moved to a different file where I put the copyright as mine; I don't really give a hoot about that so if you decide you'd prefer it as Oxide Computer then I'll happily change it.

bnaecker commented on usdt-impl/src/stap3.rs:
```diff
@@ -0,0 +1,184 @@
+            .iter()
+            .enumerate()
+            .map(|(reg_index, typ)| {
+                // Argument format is Nf@OP, N is -?{1,2,4,8} for sign and bit
+                // width, f is for floats, @ is a separator, and OP is the
+                // "actual assembly operand".
+                format!("{}@{}", typ.to_asm_size(), typ.to_asm_op(reg_index as u8))
```
Is there any documentation that describes the ABI for these STap probe records? Either for this argument description format, or the semaphore or actual probe assembler contents?

aapoalas replied:
Added documentation links to https://sourceware.org/systemtap/wiki/UserSpaceProbeImplementation mostly.

bnaecker reviewed (commented):
Thanks for taking this on, we're certainly excited about adding a new platform!

Most of the code looks pretty reasonable to me. I have a few questions about documentation and other small suggestions, but the big thing is testing. You asked about the "does it work" test -- we should definitely try to replicate that on Linux. That test works by calling out to the `dtrace` command-line utility, and checking that the probes actually fire with the expected data. Is there a similar CLI tool for SystemTap?

aapoalas commented:
> 2. I think something in the eBPF universe would make sense? Something that a typical user might employ. Do you have suggestions?

I will have to look into this a bit. I'd like to use just readelf but it might be that this will need to be SystemTap. The reason being that `does_it_work` relies on the `register_probes` method working. On Linux what one would usually do is use some CLI command to scan probes out of an executable. Also, as is tradition, at least I am not aware of there being a singular registry of USDT probes and instead each program might well have its own. If that is the case, then my guess would be that SystemTap is the one most likely to have a `/dev/dtrace/helper` equivalent somewhere.

If no equivalent filesystem based API for registering the probes can be found, then the Linux-equivalent test needs to be written in an entirely different way where it depends its own executable path... Though actually now that I think about it, that's probably not too bad? Anyway, I'll look into this.

> 4. I think stap3 makes as much sense as anything. Can you walk through considerations?

So I've been using [this official document](https://sourceware.org/systemtap/wiki/UserSpaceProbeImplementation) and [this header-only implementation](https://github.com/libbpf/usdt/blob/46c09f9440257639afc9f8cbf9dcb51f01dd9929/usdt.h#L356-L385) as my sources. `stap3` was my first idea based on the official documentation since it mentions "Only version 3 is described here." And of course "stap" comes from "SystemTap". Thinking on this now, I might prefer changing this to `stapsdt` so that it better matches readelf output (`NT_STAPSDT`): The version number is a minor detail that I doubt is very widely thought about, given that the version 3 [feature request](https://sourceware.org/bugzilla/show_bug.cgi?id=11916) was closed as done in 2010.

`systemtap` would of course be more self-explanatory and have better SEO, for the little that is worth. But, this might also lead people astray into thinking that this won't work with eBPF.

`ebpf` is a "jump on the bandwagon" kind of name. Most Linux tracing seems to prefer eBPF nowadays (when not using plain perf) and this would show the crate to be part of that movement. But, this of course also leads astray by excluding SystemTap and perf.

`linux` would probably be a misleading catch-all. I'd have to dig into the dtrace4linux code to find out what USDT format they use, but I assume there the format is DTrace with minor changes to make the ELF sections work on Linux. If that is the case, then this would obviously be misleading as the correct format for Linux would depend on whether you're using DTrace or perf/SystemTap/eBPF. If dtrace4linux (and Oracle's DTrace on Linux) all use the SystemTap format then this would of course be a perfect option.

**Personally:** I think I'm leaning into `stapsdt`. It's... not self-explanatory, but it doesn't particularly need to be. And it doesn't lead astray in any way.

>5.  I see: so with the stap implementation always performs a load from that mutable static? This seems.. nuts? Can I possibly be understanding this properly??

It does seem a bit nutty, doesn't it? But this is my understanding of the STAPSDT probes. The [documentation](https://sourceware.org/systemtap/wiki/UserSpaceProbeImplementation) mentions:

> Semaphores are treated as a counter; your tool should increment the semaphore to enable it, and decrement the semaphore when finished.

A counter in the instruction stream sounds like a bad idea, so this basically has to be in the data. And of course it can't be on the stack so either the heap or statics it is. Between those two I know I'd pick statics any day. The [header-only implementation seems to agree](https://github.com/libbpf/usdt/blob/46c09f9440257639afc9f8cbf9dcb51f01dd9929/usdt.h#L192-L196). [Another Rust impl](https://github.com/cuviper/probe-rs/blob/master/src/platform/systemtap.rs#L70-L76) actually directly uses a `static mut` per call-site, though this [causes problems](https://github.com/bpftrace/bpftrace/issues/3601).

> Are there other crates people use for Linux / eBPF / Stap / etc?

There are at least [sonde](https://github.com/wasmerio/sonde-rs) and [probe](https://github.com/cuviper/probe-rs). Sonde builds a C library that exposes functions that call the actual probes, and calls those functions through FFI from Rust. This is a perfectly working solution and probably gets compiled down to reasonable stuff with LTO or at least PGO, but I'm not a big fan of having to cross the FFI boundary there. Probe generates STAPSDT probes directly the same way usdt does, but the argument handling at least seems to be pretty bare-bones.

aapoalas reviewed (commented):
Thank you for the comments! I've addressed them now.

aapoalas commented:
> You asked about the "does it work" test -- we should definitely try to replicate that on Linux. That test works by calling out to the `dtrace` command-line utility, and checking that the probes actually fire with the expected data. Is there a similar CLI tool for SystemTap?

I was unable to find an easy answer to this: For one, the test is taking advantage of the `register_probes` function writing directly into DTrace's helper file to register the probes. I looked into Linux perf's `perf buildid-cache` command and it has something similar but it's not a single file but has at least a few parameters. I also tried looking into SystemTap itself but was unable to find anything that would directly point towards such a registering API existing.

Of course, even if such a method to register USDTs exists and is shared across all consumers on Linux (SystempTap, perf, eBPF, ...), I'd still need to figure out the data format and do the writing, which I expect would be a fair bit of work as well. Given that eg. [Brendan Gregg's Linux perf USDT notes](https://www.brendangregg.com/perf.html) starts with calling `perf buildid-cache`, I doubt there is really a feasible way to have the program itself register its own probes the way `register_probes` does.

As such, I implemented the `does_it_work` test using `readelf -n current_exe`. It wasn't too bad to do. I'll need to add cfg's to run the test on Linux and skip it otherwise. Right now I just commented out the original test. And then I need to fix all the _other_ tests that do something similar.

aapoalas commented:
Reminder: I'm waiting for review on this, specifically concerning the review fixes, if going forward with using readelf looks like it make sense to you, and how specializing the tests for Linux should be done.

andreimatei commented:
Please excuse the drive-by comment from someone trying to understand the USDT ecosystem. 

> If dtrace4linux (and Oracle's DTrace on Linux) all use the SystemTap format then this would of course be a perfect option.

Oracle's DTrace for Linux does not support the SystemTap format (i.e. the `.note.stapsdt` ELF section), does it? Instead, it supports the DTrace Object Format, right?

aapoalas commented:
> Please excuse the drive-by comment from someone trying to understand the USDT ecosystem.

At least from my point of view, none of that. Thanks for the comment.

> Oracle's DTrace for Linux does not support the SystemTap format (i.e. the `.note.stapsdt` ELF section), does it? Instead, it supports the DTrace Object Format, right?

I'm not sure, but that is probably exactly correct. Linux as usual is then plagued by two competing standards :)

aapoalas commented:
@ahl Sorry for the long wait; I have finally gotten all tests to pass on Linux! I used bpftrace to observe the probes firing: this required calling bpftrace a little differently compared to DTrace (which shouldn't be a surprise of course; no two APIs can ever be equivalent). Most notably, I had to hook into the running program using the `-p PID` flag in bpftrace.

If you could review this, that'd be swell <3

One consideration regarding the stap probes is whether this should actually be put behind a feature flag, and if so then should Linux default to stapsdt probes, or noop probes?

But in general, I think this should be good; tests pass, I was able to use this to debug a simple test program using bpftrace, and previously I've also used it to debug / trace a much more significant project.

Cheers

ahl reviewed (approved):
I mean... seems good? It works?

aapoalas commented on tests/does-it-work/src/main.rs:
```diff
@@ -38,86 +38,218 @@ fn run_test(rx: std::sync::mpsc::Receiver<()>) {
-        thr.join().expect("Failed to join test runner thread");
+
+    #[cfg(not(target_os = "linux"))]
+    mod unix {
```
Good idea <3 Fixed.

aapoalas commented on usdt-impl/src/common.rs:
```diff
@@ -110,7 +110,7 @@ pub fn construct_probe_args(types: &[DataType]) -> (TokenStream, TokenStream) {
     let abi_regs = ["rdi", "rsi", "rdx", "rcx", "r8", "r9"];
     #[cfg(target_arch = "aarch64")]
-    let abi_regs = ["x0", "x1", "x2", "x3", "x4", "x5"];
+    let arbi_regs = ["x0", "x1", "x2", "x3", "x4", "x5"];
```
Some sort of typo / random button mash  :D Fixed

aapoalas commented:
> Can we test this is CI as we do for macos and illumos?

I set up the CI to run most tests on Ubuntu as well now. I _hope_ it's correct but luckily CI will tell us that :)

aapoalas commented on .github/workflows/rust.yml:
```diff
@@ -2,9 +2,9 @@ name: Rust
 on:
   push:
-    branches: [ master ]
+    branches: [master]
```
**note:** Automatic formatter changes; I can revert these if preferred.

ahl replied:
please do

ahl commented on tests/test-json/src/main.rs:
```diff
@@ -176,29 +123,168 @@ mod tests {
                 _ => {}
             }
-            println!("DTrace not yet ready");
+            // println!("DTrace not yet ready");
```
looks like this got commented out; are there other changes?

aapoalas replied:
Hmm, this seems like an oversight/mistake; I'll fix this and go over the code once more to make sure I've not left other such unrelated changes in.

ahl commented on tests/does-it-work/src/main.rs:
```diff
@@ -38,86 +38,218 @@ fn run_test(rx: std::sync::mpsc::Receiver<()>) {
+        use std::thread;
+
+        #[test]
+        fn test_does_it_work_stapsdt() {
```
this is pretty different than the DTrace version; can you explain?

aapoalas replied:
Yeah, I originally wrote this using just readelf; I'd like to say I didn't want to dictate a DTrace alternative on Linux CI but honestly I just hadn't used them and thus went with the tool I knew.

I'll rewrite this to use bpftrace to make it match better.

aapoalas replied:
Okay, turns out I might have had more of a thought here than I remembered: bpftrace can output USDTs but will not output their arguments even with the verbose flag. Thus, I left the readelf usage and added a note about this into the test. I also renamed the test to be consistent and simplified its implementation a little to not undeservedly split up statically knowable parts (like checking "Provider:" and "does__it" strings separate instead of just checking for "Provider: does__it").

ahl commented on tests/test-json/src/main.rs:
```diff
@@ -176,29 +123,168 @@ mod tests {
+                .spawn()
+                .expect("Failed to spawn bpftrace subprocess");
+
+            // Wait for DTrace to correctly start up before notifying the test thread to start
+            // firing probes.
```
DTrace?

aapoalas replied:
Nice catch, will fix <3

bnaecker commented on usdt-impl/src/stapsdt.rs:
```diff
@@ -0,0 +1,241 @@
+///   const {
+///     // Build time: Define this location as being the place in the
+///     // instruction stream where our probe is located.
+///     core::usdt::define::<(u64, u32, *const u8)>("provider", "probe", &__usdt_sema_provider_probe);
```
The `core::` prefix is a bit misleading. This is a stand-in for the inline assembly below, right? If so, I might remove that.

aapoalas replied:
Removed the `core::` prefix and added a bit more comments around here to hopefully make the intent of this pseudo-code a little bit clearer yet.

bnaecker commented on usdt-impl/src/stapsdt/args.rs:
```diff
@@ -0,0 +1,239 @@
+    format!(
+        "{}@{}",
+        data_type_to_arg_size(typ),
+        data_type_to_asm_op(typ, reg_index as u8)
```
I'd prefer `u8::try_from(reg_index).unwrap()` here. It's unlikely we'll have > 256 arguments, but I prefer the safety anyway.

aapoalas replied:
Good point, fixed <3

bnaecker reviewed (commented):
This looks really great, thank you so much for your patience and hard work getting this integrated! I have a few small nits, but overall this looks good. The tests you've added are great.

aapoalas commented on usdt-impl/src/stapsdt.rs:
```diff
@@ -0,0 +1,241 @@
+    let impl_block = quote! {
+        {
+            #[repr(C)]
+            struct UsdtSema {
```
I'll add a comment about this but basically I'm being overly cautious here: A lot of my learning on how STAPSDTs are done come from this one single-header C library for USDTs and there they [use a C struct for this](https://github.com/libbpf/usdt/blob/46c09f9440257639afc9f8cbf9dcb51f01dd9929/usdt.h#L147). Just in case it could affect anything about anything, I thus decided to also use a `repr(C)` struct. I believe it should be absolutely okay to expose it as a `u16` directly but I was nervous about it since on some platforms a C struct like this will indeed have an effect.

aapoalas replied:
Nevermind; I just removed it. Tests (locally) still passed so the overabundance of caution is probably unneeded. (Cue some obscure ELF/compiler magic where structs are actually guaranteed to be 32-bit aligned and STAPSDT semaphores rely on that feature.)

aapoalas replied:
I also went and added an extra comment (and asked in #dark-arts though no responses yet) on it just in case this _does_ lead to something odd and someone needs to come looking for breadcrumbs.

aapoalas commented on usdt-impl/src/stapsdt/args.rs:
```diff
@@ -0,0 +1,239 @@
+    // See common.rs for note on argument passing and maximum supported
+    // argument count.
+    if reg_index > 5 {
+        unreachable!("Up to 6 probe arguments are currently supported");
```
You're goddamn right! Fixed <3

aapoalas commented:
> This looks really great, thank you so much for your patience and hard work getting this integrated! I have a few small nits, but overall this looks good. The tests you've added are great.

Addressed all the comments, thank you for them <3 And thank you and ahl for the patience as well <3
